### DISCUSSION TITLE
Adding a new fluid

### DISCUSSION THREAD

**Discussion by theo-bech** [oemof/tespy#407] [Q&A]
Hi all, 

I'm new to TESpy and I've been researching how I can add a fluid that's not part of the existing catalogue for my simulation. Can I manually input the fluid properties vs. temperature somewhere? I read in a previous discussion that this might not be possible, but I wasn't entirely sure the issue was the same, as additional properties seemed to be the focus. 

Thank you!

**Comment by fwitte** [comment]
Dear @theo-bech,

TL;DR: There are very limited and rather hacky options at the moment.

- With the current architecture you basically have to create some custom functions and replace inbuilt functions in you localy copy of TESPy to do that. I did that for some polynomes to validate a model on this branch: https://github.com/oemof/tespy/tree/features/KKH_combustion_validation/src/tespy/tools. You can have a look at `kkh.py` and the respective implementation in the `fluid_properties.py` module. However it is not really stable and mostly untested, the only use case is this one: https://github.com/KarimHShawky/Chemical-Exergy-in-TESPy. If you have some time this might be a possiblilty.
- Another option would be to work on the new design for a new fluid property back end I had started to work with a couple months ago. This would make it easy to include custom fluid property functions https://github.com/oemof/tespy/tree/features/flexible-fluid-property-back-end/fluid_properties, but did not have the time to finish that. It needs quite some work for implementing the new design in the software. If you are willing to input some more time on this, we could have some discussion about how the solution should look like and how to implement it.
- Last option is to build own CoolProp fluids, there are some information on how to do that on their website, also maybe consider this issue https://github.com/CoolProp/CoolProp/issues/1931. I have never done something like this, so I would not really be of help here.

Best

Francesco

**Reply by Jesus-Fuentes** [reply]
Hello,

About this, am I right to think that as of now Tespy does not deal with Humid Air past the saturation point? 
I've seen your discussion under '[Air compressor simulation inconsistency · Issue #206]'.  Tespy cannot deal with vapour in the air condensing, am I right?

The concept you described on 4 Jun 2020:
_For every gas component check, whether the temperature of the mixture is below critical temperature of the gas component. If so
Calculate the partial pressure and the saturation pressure of the gas component.
If the partial pressure is higher than the saturation pressure, calculate gaseous state at saturation pressure and calculate liquid state at pressure of the stream._
has not been implemented, right?

**Reply by fwitte** [reply]
@Jesus-Fuentes,

TESPy does deal with condensation of water in gases, it had been implemented sometime in the last year if I recall correctly. Do you have an example, where you have doubts on this?

Also, there is currently work on reimplementing the whole fluid property back-end at #384, which will allow users to select the way they want to handle the fluid properties of such mixtures. This will take a couple more weeks however.

Best

Francesco

**Reply by fwitte** [reply]
And to follow up @theo-bech: Incase you are still considering using TESPy, the feature you need will be available with #384 as well :).

**Reply by Jesus-Fuentes** [reply]
Thanks very much for the prompt reply. 

I'm pretty new to Tespy so I'm just starting with an exhaust from a fuel cell that is basically humid air and putting it through a heat exchanger to cool it down.

After reading Issue #206 I was trying to get confirmation that the heat exchanger (the basic one, not the simple, not the condenser) will be able to calculate amount of vapour being condensed.

Is humid air still defined the way it was in Issue #206 (mixture of 'air', 'water') or is there a 'humidair' in the database?

**Reply by Jesus-Fuentes** [reply]
```python
from tespy.networks import Network
from tespy.connections import Connection
from tespy.components import Source
from tespy.components import Sink
from tespy.components import HeatExchangerSimple


# define cycle
cycle = Network(fluids=['air', 'water'])
cycle.set_attr(p_unit='bar', T_unit='C', v_unit='m3 / h')
cycle.set_attr(iterinfo=False)

# define components
inlet = Source('Inlet')
HE = HeatExchangerSimple('HE')
outlet = Sink('Outlet')

# define connections
conn1 = Connection(inlet, 'out1', HE, 'in1')
conn2 = Connection(HE, 'out1', outlet, 'in1')

conn1.set_attr(p=2, T=170, v=1, fluid={'water': 0.5, 'air': 0.5})

cycle.add_conns(conn1, conn2)

HE.set_attr(pr=0.99, design=['pr'], offdesign=['zeta'])
# %% key parameter
HE.set_attr(Q=-33.05)

cycle.solve('design')
cycle.print_results()
```

**Reply by Jesus-Fuentes** [reply]
That was meant to look like code - it looks awful in my browser, apologies if it does on yours too.

It's very basic as you can see. Flow from a source to a sink via a simple heat exchanger where I play with the amount of heat removed. The flow is a mixture of water and air at 2 bar, 170 deg C.

When Q reaches -33.05 (not sure about the units to be fair), the outlet T has dropped to 105 deg C already. 

```
##### RESULTS (HeatExchangerSimple) #####
+----+-----------+----------+----------+-----+-----+------+------+--------+
|    |         Q |       pr |     zeta |   D |   L |   ks |   kA |   Tamb |
|----+-----------+----------+----------+-----+-----+------+------+--------|
| HE | -3.31e+01 | 9.90e-01 | 2.84e+10 | nan | nan |  nan |  nan |    nan |
+----+-----------+----------+----------+-----+-----+------+------+--------+
##### RESULTS (Connection) #####
+--------------------+-----------+-----------+-----------+-----------+
|                    |         m |         p |         h |         T |
|--------------------+-----------+-----------+-----------+-----------|
| Inlet:out1_HE:in1  | 3.363e-04 | 2.000e+00 | 1.693e+06 | 1.700e+02 |
| HE:out1_Outlet:in1 | 3.363e-04 | 1.980e+00 | 1.595e+06 | 1.053e+02 |
+--------------------+-----------+-----------+-----------+-----------+
```

If I so much increase Q to -33.06 then I get the following

Could not find a feasible value for mixture temperature at connection HE:out1_Outlet:in1. The values for temperature, specific volume, volumetric flow and entropy are set to nan.

```
##### RESULTS (HeatExchangerSimple) #####
+----+-----------+----------+--------+-----+-----+------+------+--------+
|    |         Q |       pr |   zeta |   D |   L |   ks |   kA |   Tamb |
|----+-----------+----------+--------+-----+-----+------+------+--------|
| HE | -3.31e+01 | 9.90e-01 |    nan | nan | nan |  nan |  nan |    nan |
+----+-----------+----------+--------+-----+-----+------+------+--------+
##### RESULTS (Connection) #####
+--------------------+-----------+-----------+-----------+-------------+
|                    |         m |         p |         h |           T |
|--------------------+-----------+-----------+-----------+-------------|
| Inlet:out1_HE:in1  | 3.363e-04 | 2.000e+00 | 1.693e+06 |   1.700e+02 |
| HE:out1_Outlet:in1 | 3.363e-04 | 1.980e+00 | 1.595e+06 | nan         |
+--------------------+-----------+-----------+-----------+-------------+
```

That makes me think the vapour condensation is not being handled?

Incidentally, I believe the boiling point at 1.98 bars (outlet P) is something like 115-120 deg C, so not exactly the 105 deg C 'cut off' I'm getting.

Is it me defining the problem poorly?

**Reply by fwitte** [reply]
@Jesus-Fuentes:

You can syntax highlight your code like this:

````
```python
(here is your code)
```
````

To your question: Your massflow might be too low for the precision of TESPy's solver, it is around `0.001`. A volumetric flow of 1 m3/h is really too small I fear.

I checked with the following example, and it works as intended. I reduce the temperature step by step (numerical difficulties for the solver). You also experience, that from around 105 °C the solver get slower. The reason for that is, that the water starts to condense at that temperature (the partial (not the total pressure!) pressure of the water is near saturation), which is computationally more expensive since the saturation pressure has to be updated in every iteration.

```python
from tespy.networks import Network
from tespy.connections import Connection
from tespy.components import Source
from tespy.components import Sink
from tespy.components import HeatExchangerSimple


# define cycle
cycle = Network(fluids=['air', 'water'])
cycle.set_attr(p_unit='bar', T_unit='C', v_unit='m3 / h')
cycle.set_attr(iterinfo=False)

# define components
inlet = Source('Inlet')
HE = HeatExchangerSimple('HE')
outlet = Sink('Outlet')

# define connections
conn1 = Connection(inlet, 'out1', HE, 'in1')
conn2 = Connection(HE, 'out1', outlet, 'in1')

conn1.set_attr(p=2, T=170, m=1, fluid={'water': 0.5, 'air': 0.5})
conn2.set_attr(T=110)

cycle.add_conns(conn1, conn2)

HE.set_attr(pr=0.99)
# %% key parameter
HE.set_attr()

cycle.solve('design')
cycle.print_results()

for T in range(70, 110)[::-1]:
    conn2.set_attr(T=T)

    cycle.solve('design')
    cycle.print_results()
```

**Reply by Jesus-Fuentes** [reply]
Thanks very much. I appreciate the time you've taken to check this.

I wanted to double check something else, please.

I've tried to do the process manually in Excel with basic stuff: steam tables, and cp 1.0035 KJ/Kg/K for dry air enthalpy.

I've divided the cooling process in half a degree drops, all the way from 170 to 70 deg C.

Only mod to your code is that I've set up these units: h_unit='kJ / kg', m_unit='kg / h' to make sure.
m=1 and 50% air mass 50% water, as in your code.

I've calculated, when condensation starts, the amount of water that goes from vapour to liquid in each step i.
Then for every step (-0.5 deg C drop like I said) I've listed:
- The drop in dry air enthalpy from step i-1 to i
- The drop in vapour enthalpy for the vapour that remains from step i-1 to i
- The condensation enthalpy for the vapour that condenses from step i-1 to i
- The drop in liquid enthalpy for the liquid that already existed in step i-1

This way I get a Delta Enthalpy for my full 170->70 process of -1151 KJ/kg.
Your last iteration yields

h
1.693e+03 
5.171e+02

that is, -1176 KJ/kg drop.

I'm sure your model is far more detailed so I guess this is good enough?


My question is about the Q reported, please. Tespy shows a Q of -3.27e+02. How does that relate to the h values of 1.693e+03, 5.171e+02, given that m=1? Is it a matter of units? What units does Q come in? The ones you defined with 'h_unit' ?

**Reply by fwitte** [reply]
> I'm sure your model is far more detailed so I guess this is good enough?

It is not that complicated, just calling CoolProp pure fluids functions and applying a mixture routine.

> My question is about the Q reported, please. Tespy shows a Q of -3.27e+02.

If I run this on my machine I get -1.176 MW as a result. The equation for heat is just mass flow times change of enthalpy.

> My question is about the Q reported, please. Tespy shows a Q of -3.27e+02.

The units of all component parameters are always SI units, a proper unit system could or should be implemented here, but I did not have the time for that.

**Reply by Jesus-Fuentes** [reply]
Thanks very much. Sorry to insist, but how is the -3.27e+02 derived then? I'm playing with 1176 and 327 and can't connect them via the mass flow (m=1)!

**Reply by fwitte** [reply]
> Thanks very much. Sorry to insist, but how is the -3.27e+02 derived then? I'm playing with 1176 and 327 and can't connect them via the mass flow (m=1)!

I struggle to understand, what you want to say here. For the lowest temperature value I do get the correct results, I do not see where the 327 comes from, what are your inputs for that result?

**Reply by Jesus-Fuentes** [reply]
```python
from tespy.networks import Network
from tespy.connections import Connection
from tespy.components import Source
from tespy.components import Sink
from tespy.components import HeatExchangerSimple


# define cycle
cycle = Network(fluids=['air', 'water'])
cycle.set_attr(p_unit='bar', T_unit='C', v_unit='m3 / h', h_unit='kJ / kg',
               m_unit='kg / h')
cycle.set_attr(iterinfo=False)

# define components
inlet = Source('Inlet')
HE = HeatExchangerSimple('HE')
outlet = Sink('Outlet')

# define connections
conn1 = Connection(inlet, 'out1', HE, 'in1')
conn2 = Connection(HE, 'out1', outlet, 'in1')

conn1.set_attr(p=2, T=170, m=1, fluid={'water': 0.5, 'air': 0.5})
conn2.set_attr(T=110)

cycle.add_conns(conn1, conn2)

HE.set_attr(pr=0.99)
# %% key parameter
HE.set_attr()

cycle.solve('design')
cycle.print_results()

for T in range(70, 110)[::-1]:
    conn2.set_attr(T=T)

    cycle.solve('design')
    cycle.print_results()
```

That's the code I've run. And the final output comes out as this:

```
##### RESULTS (HeatExchangerSimple) #####
+----+-----------+----------+----------+-----+-----+------+------+--------+
|    |         Q |       pr |     zeta |   D |   L |   ks |   kA |   Tamb |
|----+-----------+----------+----------+-----+-----+------+------+--------|
| HE | -3.27e+02 | 9.90e-01 | 7.73e+10 | nan | nan |  nan |  nan |    nan |
+----+-----------+----------+----------+-----+-----+------+------+--------+
##### RESULTS (Connection) #####
+--------------------+-----------+-----------+-----------+-----------+
|                    |         m |         p |         h |         T |
|--------------------+-----------+-----------+-----------+-----------|
| Inlet:out1_HE:in1  | 1.000e+00 | 2.000e+00 | 1.693e+03 | 1.700e+02 |
| HE:out1_Outlet:in1 | 1.000e+00 | 1.980e+00 | 5.171e+02 | 7.000e+01 |
+--------------------+-----------+-----------+-----------+-----------+
```

**Reply by Jesus-Fuentes** [reply]
See the Q -3.27e+02 ?

**Reply by fwitte** [reply]
It is because you have the mass flow unit `kg / h`. So 1 kg / h * 1 h / 3600 s * 1176 kJ / kg = 0.0327 kW or 327 W.

**Reply by Jesus-Fuentes** [reply]
Ahh! I had 1176/327 = 3.596 but the penny didn't drop!
Thanks. One last question, please. The code so far has no way of showing the mass of water that has condensed? 

I've tried  having a peek at this:

```dir(conn2)```

but I haven't found anything. This gives you the composition indeed, but no further info on water state:

```conn2.fluid.val```
OrderedDict([('air', 0.49999999999999994), ('water', 0.49999999999999994)])

**Comment by fwitte** [comment]
@theo-bech: If still relevant, the new version of tespy opens its API to implement user specified fluids, see https://tespy.readthedocs.io/en/main/modules/fluid_properties.html#tespy-fluid-properties-label and https://github.com/oemof/tespy/tree/dev/src/tespy/tools/fluid_properties
